### PR TITLE
test: Add feature_taproot case involving invalid internal pubkey

### PR DESCRIPTION
**Pull request by sipa**
Add a test case to feature_taproot which involves an output that is (incorrectly) constructed, using an invalid internal public key and valid script tree. It is designed to detect cases where the script path spending validation logic does not detect this case, and instead treats the internal public key as the point at infinity.

Equivalent unit test case added in https://github.com/bitcoin-core/qa-assets/pull/98.

### CODE REVIEW

File: test/functional/test_framework/script.py [hunk 1]
```diff
@@ -872,7 +872,7 @@ def taproot_tree_helper(scripts):
 # - merklebranch: the merkle branch to use for this leaf (32*N bytes)
 TaprootLeafInfo = namedtuple("TaprootLeafInfo", "script,version,merklebranch,leaf_hash")
 
-def taproot_construct(pubkey, scripts=None):
+def taproot_construct(pubkey, scripts=None, treat_internal_as_infinity=False):
```
Comment (instagibbs) [inline test/functional/test_framework/script.py:875]: needs to be described in below comments

File: test/functional/feature_taproot.py [hunk 2]
```diff
@@ -661,6 +668,23 @@ def spenders_taproot_active():
...
+    # == Test involving an internal public key not on the curve ==
+
+    # Generate an invalid public key
+    while True:
+        invalid_pub = random_bytes(32)
+        if not SECP256K1.is_x_coord(int.from_bytes(invalid_pub, 'big')):
+            break
+
+    # Implement a test case that detects validation logic which maps invalid public keys to the
+    # point at infinity in the tweaking logic.
```
Comment (instagibbs) [inline test/functional/feature_taproot.py:680]: can this be dumbed down for those of us who treat EC things in a black box manner? It's not on the curve, what does treating it as a point at infinity mean in layman?

File: test/functional/test_framework/script.py (hunk 1 above)
Comment (instagibbs) [reply test/functional/test_framework/script.py:875]: > if they don't verify whether or not a revealed internal public key in a script path spend is valid, and instead implement output_key == tweak(internal_key, tweakval) but such that tweak(invalid_key, tweakval) equals the public key corresponding to private key tweakval

Ok so you're saying an implementation *could*, instead of fail detecting it's not on the curve, could instead just consider the pubkey as the point at infinity, which would result in the tweak alone being the key, and in fact this may be a common mistake.

**Review by aureleoules — COMMENTED** [review]
ACK 6ca24638d0b70f3d40166ca1894f2481260a1b5f
I understand that an invalid x-only pubkey can be treated as a point to infinity as both lead to invalid results.
This is only testing the test framework though correct? Which is still useful.

**Comment by instagibbs** [comment]
ACK 5d413c8e793a439540d064d24fddfc868e1817d0

**Review by aureleoules — APPROVED** [review]
reACK 5d413c8e793a439540d064d24fddfc868e1817d0
